### DISCUSSION TITLE
Possible to launch a Dev Branch of Raspotify that Tracks Dev Branch of Librespot

### DISCUSSION THREAD

**Discussion by txoof** [dtcooper/raspotify#626] [Q&A]
I have a project ([PaperPi](https://github.com/txoof/PaperPi)) that depends on SpoCon/librespot-java which has a built-in ability to pull now-playing information and album art. 

Something has gone off the rails with regards to the Java version and the latest version of Bullseye. [Local builds fail](https://github.com/librespot-org/librespot-java/issues/613) and SpoCon (which provides a tidy debian package) [crashes on play.](https://github.com/spocon/spocon/issues/55)

After some coaching from @JasonLG1979, I built the [dev](https://github.com/librespot-org/librespot/tree/dev) branch of librespot and used the onevent sample script to get pretty much everything I need.

I've installed raspotify and then just plunked in the newly built dev version of librespot into `/usr/bin/` and tested it out. It works perfect for my needs, but it is rather a round-about way of doing this. The one caveat is that the config file needs to have an explicity `LIBRESPOT_AUTOPLAY=on` directive for the service to start properly.

Is there a possibility to build raspotify-dev against librespot-dev?

**Comment by JasonLG1979** [comment]
You could just fork Raspotify and point to the librespot dev branch in the build script and make whatever changes you need elsewhere.

I have no plans on doing Raspotify-dev

**Reply by txoof** [reply]
Fair.

Are there instructions for setting up a development environment. I see in the build script that it requires Docker. I have zero experience with docker and not really sure where to begin.  Does [this look like a legit place to start](https://www.simplilearn.com/tutorials/docker-tutorial/raspberry-pi-docker)?

**Reply by JasonLG1979** [reply]
Sure, the packages are built in a docker container so you can basically build it on anything:

https://github.com/dtcooper/raspotify/wiki/Building-the-Package(s)-Yourself

And more detailed instructions for compiling librespot binaries you already found.

**Reply by txoof** [reply]
This is super helpful! I've got docker installed and I've been trying to create a docker environment based on the Dockerfile. 

I don't have an AMD64 environment, only an M1 and Pi4. I'm trying to set up the Docker environment to build using `docker build -t raspotify .` It fails with the error below. To remedy this, does it make sense to remove the references aarch64 and arm64 from `Dockerfile`?

[edit] I should probably mention that I'm only interested in building the armhf version for the moment.

```
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 build-essential : Depends: gcc (>= 4:10.2) but it is not installable
                   Depends: g++ (>= 4:10.2) but it is not installable
 dpkg-dev : Depends: binutils but it is not installable
 g++-10-arm-linux-gnueabihf:arm64 : Depends: libstdc++-10-dev-armhf-cross:arm64 (>= 10) but it is not installable
 gcc-10-arm-linux-gnueabihf:arm64 : Depends: binutils-arm-linux-gnueabihf:arm64 (>= 2.35.1) but it is not going to be installed
                                    Depends: libgcc-10-dev-armhf-cross:arm64 (>= 10.2.1-6cross1) but it is not installable
E: Unable to correct problems, you have held broken packages.
The command '/bin/sh -c mkdir /build     && mkdir /.cargo     && rustup target add aarch64-unknown-linux-gnu     && rustup target add armv7-unknown-linux-gnueabihf     && echo '[target.aarch64-unknown-linux-gnu]\nlinker = "aarch64-linux-gnu-gcc"' > /.cargo/config     && echo '[target.armv7-unknown-linux-gnueabihf]\nlinker = "arm-linux-gnueabihf-gcc"' >> /.cargo/config     && cargo install --jobs "$(nproc)" cargo-deb     && dpkg --add-architecture arm64     && dpkg --add-architecture armhf     && apt-get update     && apt-get -y upgrade     && apt-get install -y --no-install-recommends         build-essential         libasound2-dev         libpulse-dev         crossbuild-essential-arm64         libasound2-dev:arm64         libpulse-dev:arm64         crossbuild-essential-armhf         libasound2-dev:armhf         libpulse-dev:armhf         git         bc         dpkg-dev         liblzma-dev         pkg-config         gettext-base     && rm -rf /var/lib/apt/lists/*     && git config --global --add safe.directory /mnt/raspotify' returned a non-zero code: 100
```

**Reply by JasonLG1979** [reply]
You might be able to get docker, git and make installed on the M1? Idk? As far as the Pi4 I'd try a sudo apt update && sudo apt dist-upgrade

**Reply by JasonLG1979** [reply]
I bounce back and forth between Fedora and Ubuntu on my main machine and I run Debian on my server so those are the only environments I can say anything will work on. I've never installed docker on my pi zero v2 (the only pi I have) I mainly just use it as a performance checker when I'm writing code. Less powerful CPUs are better at showing differences compared to 16 core beasts that librespot is barely a blip on.

**Reply by txoof** [reply]
Thanks for your patience. I've learned a ton through this exercise. 

I also think I'm starting to grok how this all fits together. Can you check my understanding and point out what I'm missing? Following these steps seem to be working as I type this.

I think I might be making some redundant steps (4, 5 below).

## Prerequsites

1.  Docker needs to be installed and available for the currently signed in user 
2.  `git` and `build-essential` need to be installed 
3.  raspotify repo is cloned locally
4.  Remove references to AMD and ARM64 from `Dockerfile` to prevent Docker from pulling in unneeded packages and failing on Pi
    - Unsure if this is necessary
5. Docker container needs to exist and be built locally using `docker build -t raspotify .` 
   - Unsure if this is necessary 

## Building the armhf version:

- `make armhf` 
- Kick back and do some knitting/chainsaw sculpting while it chooches 


## To Build against Librespot's Dev Branch

- edit `build.sh` and edit line 44 to point to the dev branch at librespot

**Reply by JasonLG1979** [reply]
5 is not needed. The make file does that.

**Reply by JasonLG1979** [reply]
You will also need to change the build profile to release when building upstream librespot.

**Reply by JasonLG1979** [reply]
I usually also prune the container right after I build or at least before I rebuild to make sure that I'm building with the latest base. For me anyway the docker build environment is a single use thing in this case. I'm sure there's a way to tell docker to never save/cache it for later?

**Reply by txoof** [reply]
> You will also need to change the build profile to release when building upstream librespot.

Unsure what you mean by this?


> I usually also prune the container right after I build or at least before I rebuild to make sure that I'm building with the latest base.

So just `docker container prune` and kill the stopped containers?

After substituting in `https://github.com/librespot-org/librespot.git` into `build.sh`, I ran another make (without pruning) and everything went as expected, but kicked the error below and failed to produce a .deb file. I suspect this has something to do with lines 65-78 in `build.sh`. Any ideas what I need to move past this?

```
   Compiling librespot v0.4.2-raspotify-fork (/mnt/raspotify/librespot)
    Finished raspotify [optimized] target(s) in 12m 44s
Copy Librespot binary to package root...
'/build/armv7-unknown-linux-gnueabihf/raspotify/librespot' -> 'raspotify/usr/bin/librespot'
Generate Debian control...
Build Raspotify deb...
dpkg-deb: error: parsing file 'raspotify/DEBIAN/control' near line 2 package 'raspotify':
 'Version' field value '0.43.46~librespot.unknown-': revision number is empty
make: *** [Makefile:8: armhf] Error 2
```

**Reply by txoof** [reply]
when I try to run the commands on lines 65/66 manually from the librespot cloned repo they appear to work:
```
$ git describe --tags "$(git rev-list --tags --max-count=1)" 2>/dev/null
v0.4.2

$ git rev-parse HEAD | cut -c 1-7 2>/dev/null
fc549b9
```

Not sure why they yield `'0.43.46~librespot.unknown-'` when run from in the build script.
